File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        def inorder(node, array):
            # inorder traversal
            if node is None:
                return

            array.append(node.val)
            inorder(node.left, array)
            inorder(node.right, array)

            return

        def build(array):
            if len(array) == 0:
                return None

            left, right = 0, len(array) - 1
            mid = (left + right) // 2

            #print("len(array), mid: ", len(array), mid)  
            left_node = build(array[:mid]) 
            right_node = build(array[mid+1:]) 
            mid_node = TreeNode(array[mid], left_node, right_node)

            return mid_node

        inorder_list = []
        inorder(root, inorder_list)    
        inorder_list.sort()
        #print("inorder_list: ", inorder_list)

        return build(inorder_list)
        

File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* balanceBST(TreeNode* root) {
        vector<int> inorder;
        inorderTraverse(root, inorder);
        int size = inorder.size();
        return create(inorder, 0, size-1); 
    }
private:
    void inorderTraverse(TreeNode* node, vector<int>& inorder) {
        if (node == nullptr) return;
        inorderTraverse(node->left, inorder);
        inorder.push_back(node->val);
        inorderTraverse(node->right, inorder);
    }

    TreeNode* create(const vector<int>& inorder, int start, int end) {
        if (start > end) return nullptr;
        int mid = (start + end) / 2;
        TreeNode* left = create(inorder, start, mid - 1);
        TreeNode* right = create(inorder, mid + 1, end);
        return new TreeNode(inorder[mid], left, right);
    }
};